impeller/entity/geometry/stroke_path_geometry.cc

```diff
@@ -38,225 +82,312 @@ Join StrokePathGeometry::GetStrokeJoin() const {
+        // contours with two zero volume triangles, which will be discarded by
+        // the rasterizer).
+        vtx.position = polyline.GetPoint(contour_start_point_i - 1);
+        // Append two vertices when "picking up" the pen so that the triangle
+        // drawn when moving to the beginning of the new contour will have zero
+        // volume.
+        vtx_builder.AppendVertex(vtx.position);
+        vtx_builder.AppendVertex(vtx.position);
+
+        vtx.position = polyline.GetPoint(contour_start_point_i);
+        // Append two vertices at the beginning of the new contour, which
+        // appends  two triangles of zero area.
+        vtx_builder.AppendVertex(vtx.position);
+        vtx_builder.AppendVertex(vtx.position);
+      }
 
-  // Offset state.
-  Point offset;
-  Point previous_offset;  // Used for computing joins.
+      // Generate start cap.
+      if (!polyline.contours[contour_i].is_closed) {
+        auto cap_offset =
+            Vector2(-contour.start_direction.y, contour.start_direction.x) *
+            stroke_width * 0.5;  // Counterclockwise normal
+        cap_proc(vtx_builder, polyline.GetPoint(contour_start_point_i),
+                 cap_offset, scale, true);
+      }
+
+      for (size_t contour_component_i = 0;
+           contour_component_i < contour.components.size();
+           contour_component_i++) {
+        auto component = contour.components[contour_component_i];
+        auto is_last_component =
+            contour_component_i == contour.components.size() - 1;
+
+        auto component_start_index = component.component_start_index;
+        auto component_end_index =
+            is_last_component ? contour_end_point_i - 1
+                              : contour.components[contour_component_i + 1]
+                                    .component_start_index;
+        if (component.is_curve) {
+          AddVerticesForCurveComponent(
+              component_start_index, component_end_index, contour_start_point_i,
+              contour_end_point_i, contour);
+        } else {
+          AddVerticesForLinearComponent(
+              component_start_index, component_end_index, contour_start_point_i,
+              contour_end_point_i, contour);
+        }
+      }
+
+      // Generate end cap or join.
+      if (!contour.is_closed) {
+        auto cap_offset =
+            Vector2(-contour.end_direction.y, contour.end_direction.x) *
+            stroke_width * 0.5;  // Clockwise normal
+        cap_proc(vtx_builder, polyline.GetPoint(contour_end_point_i - 1),
+                 cap_offset, scale, false);
+      } else {
+        join_proc(vtx_builder, polyline.GetPoint(contour_start_point_i), offset,
+                  contour_first_offset, scaled_miter_limit, scale);
+      }
+    }
+  }
 
   // Computes offset by calculating the direction from point_i - 1 to point_i if
   // point_i is within `contour_start_point_i` and `contour_end_point_i`;
   // Otherwise, it uses direction from contour.
```

Comment:
nit: An offset is usually a difference between 2 points.  I'm not sure what 2 points these are from reading this.  It just talks about the direction.  Also it's not proper doxygen formatting (needs `///`).

Reply:
Updated the doc comment, but I'd like to leave the logic/naming as is.